packages/runtime/datastore/src/dataStoreRuntime.ts

```diff
@@ -851,16 +884,34 @@ IFluidDataStoreChannel, IFluidDataStoreRuntime, IFluidHandleContext {
     private attachListener() {
         this.setMaxListeners(Number.MAX_SAFE_INTEGER);
         this.dataStoreContext.once("attaching", () => {
-            assert(this.bindState !== BindState.NotBound,
-                0x186 /* "Data store attaching should not occur if it is not bound" */);
+            /**
+             * back-compat 0.58.2000 - Ideally, attachGraph() should have already been called making the data store
+             * locally visible. However, before visibility state was added, this may not have been the case and data
+             * store can move to "attaching" state in 2 scenarios:
+             * 1) Before attachGraph() is called - When a data store is created and bound in an attached container.
+             * 2) After attachGraph() is called - When a detached container is attached.
+             *
+             * The basic idea is that all local object should become locally visible before they are globally visible.
+             */
+            this.attachGraph();
+
             this._attachState = AttachState.Attaching;
+
+            assert(this.visibilityState === VisibilityState.LocallyVisible,
+                `Data store should not attach if is not locally visible - ${this.visibilityState}`);
+
+            // Mark the data store globally visible and make its child channels visible as well.
+            this.visibilityState = VisibilityState.GloballyVisible;
+            this.localChannelContextQueue.forEach((channel) => {
+                channel.makeVisible();
+            });
+            this.localChannelContextQueue.clear();
+
             // This promise resolution will be moved to attached event once we fix the scheduler.
             this.deferredAttached.resolve();
             this.emit("attaching");
         });
         this.dataStoreContext.once("attached", () => {
-            assert(this.bindState === BindState.Bound,
```

Comment:
Should we be checking the visibilityState here?

Reply:
We could assert that visibility state is globally visible. I didn't add it because we don't care about "attached" event. We should eventually remove attachState. I will create an issue to do this when (or if) this PR is merged. A lot of state will have to be cleared once there aren't back-compat concerns.